BLD: warning from `scipy/special/special/gamma.h`

There's a new warning when building main:

```
[32/247] Compiling C++ object scipy/specia...86_64-linux-gnu.so.p/_special_ufuncs.cpp.o
In file included from ../scipy/special/_special_ufuncs.cpp:15:
../scipy/special/special/gamma.h: In function 'std::complex<_Tp> special::gamma(std::complex<_Tp>) [with T = float]':
../scipy/special/special/gamma.h:19:44: warning: infinite recursion detected [-Winfinite-recursion]
   19 | SPECFUN_HOST_DEVICE inline std::complex<T> gamma(std::complex<T> z) {
      |                                            ^~~~~
../scipy/special/special/gamma.h:20:17: note: recursive call
   20 |     return gamma(z);
      |            ~~~~~^~~
```

--- a/scipy/special/special/gamma.h
+++ b/scipy/special/special/gamma.h
@@ -17,7 +17,7 @@ SPECFUN_HOST_DEVICE T gammaln(T x) {
 
 template <typename T>
 SPECFUN_HOST_DEVICE inline std::complex<T> gamma(std::complex<T> z) {
-    return gamma(z);
+    return gamma_complex(z);
 }
 
 } // namespace special

--- a/scipy/special/special/gamma.h
+++ b/scipy/special/special/gamma.h
@@ -15,9 +15,18 @@ SPECFUN_HOST_DEVICE T gammaln(T x) {
     return cephes::lgam(x);
 }
 
-template <typename T>
-SPECFUN_HOST_DEVICE inline std::complex<T> gamma(std::complex<T> z) {
-    return gamma_complex(z);
+SPECFUN_HOST_DEVICE inline std::complex<double> gamma(std::complex<double> z) {
+    // Compute Gamma(z) using loggamma.
+    if (z.real() <= 0 && z == std::floor(z.real())) {
+        // Poles
+        set_error("gamma", SF_ERROR_SINGULAR, NULL);
+        return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+    }
+    return std::exp(loggamma(z));
+}
+
+SPECFUN_HOST_DEVICE inline std::complex<float> gamma(std::complex<float> z) {
+    return static_cast<std::complex<float>>(gamma(static_cast<std::complex<double>>(z)));
 }
 
 } // namespace special

--- a/scipy/special/special/loggamma.h
+++ b/scipy/special/special/loggamma.h
@@ -143,11 +143,21 @@ SPECFUN_HOST_DEVICE inline std::complex<float> loggamma(std::complex<float> z) {
     return static_cast<std::complex<float>>(loggamma(static_cast<std::complex<double>>(z)));
 }
 
-SPECFUN_HOST_DEVICE inline std::complex<double> gamma(std::complex<double> z) {
+SPECFUN_HOST_DEVICE inline std::complex<double> gamma_complex(std::complex<double> z) {
     // Compute Gamma(z) using loggamma.
     if (z.real() <= 0 && z == std::floor(z.real())) {
         // Poles
-        set_error("gamma", SF_ERROR_SINGULAR, NULL);
+        set_error("gamma_complex", SF_ERROR_SINGULAR, NULL);
+        return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
+    }
+    return std::exp(loggamma(z));
+}
+
+SPECFUN_HOST_DEVICE inline std::complex<float> gamma_complex(std::complex<float> z) {
+    // Compute Gamma(z) using loggamma.
+    if (z.real() <= 0 && z == std::floor(z.real())) {
+        // Poles
+        set_error("gamma_complex", SF_ERROR_SINGULAR, NULL);
         return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
     }
     return std::exp(loggamma(z));

--- a/scipy/special/special/loggamma.h
+++ b/scipy/special/special/loggamma.h
@@ -143,17 +143,6 @@ SPECFUN_HOST_DEVICE inline std::complex<float> loggamma(std::complex<float> z) {
     return static_cast<std::complex<float>>(loggamma(static_cast<std::complex<double>>(z)));
 }
 
-template <typename T>
-SPECFUN_HOST_DEVICE inline std::complex<T> gamma_complex(std::complex<T> z) {
-    // Compute Gamma(z) using loggamma.
-    if (z.real() <= 0 && z == std::floor(z.real())) {
-        // Poles
-        set_error("gamma_complex", SF_ERROR_SINGULAR, NULL);
-        return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
-    }
-    return std::exp(loggamma(z));
-}
-
 SPECFUN_HOST_DEVICE inline double rgamma(double z) { return cephes::rgamma(z); }
 
 SPECFUN_HOST_DEVICE inline float rgamma(float z) { return rgamma(static_cast<double>(z)); }

--- a/scipy/special/special/loggamma.h
+++ b/scipy/special/special/loggamma.h
@@ -143,17 +143,8 @@ SPECFUN_HOST_DEVICE inline std::complex<float> loggamma(std::complex<float> z) {
     return static_cast<std::complex<float>>(loggamma(static_cast<std::complex<double>>(z)));
 }
 
-SPECFUN_HOST_DEVICE inline std::complex<double> gamma_complex(std::complex<double> z) {
-    // Compute Gamma(z) using loggamma.
-    if (z.real() <= 0 && z == std::floor(z.real())) {
-        // Poles
-        set_error("gamma_complex", SF_ERROR_SINGULAR, NULL);
-        return {std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN()};
-    }
-    return std::exp(loggamma(z));
-}
-
-SPECFUN_HOST_DEVICE inline std::complex<float> gamma_complex(std::complex<float> z) {
+template <typename T>
+SPECFUN_HOST_DEVICE inline std::complex<T> gamma_complex(std::complex<T> z) {
     // Compute Gamma(z) using loggamma.
     if (z.real() <= 0 && z == std::floor(z.real())) {
         // Poles

